hoomd/data/data_structures.py

```diff
@@ -0,0 +1,497 @@
+                rtn_type_def = type_def.converter.cond
+        elif isinstance(type_def.converter, Either):
+            for spec in type_def.converter.specs:
+                matches = []
+                if isinstance(spec, desired_type):
+                    matches.append(spec)
+                if len(matches) > 1:
+                    raise ValueError(
+                        "Parameter defined with multiple valid definitions of "
+                        "the same type.")
+                elif len(matches) == 1:
+                    rtn_type_def = matches[0]
+    # When no valid type converter can be found error out.
+    if rtn_type_def is None:
+        raise ValueError("Unable to find type definition for attribute.")
+    return rtn_type_def
+
+
+def _to_hoomd_data_structure(data, type_def, parent=None, label=None):
+    """Convert raw data to a synced _HOOMDDataStructures object.
+
+    This does nothing if the data is not detected to be a supported HOOMD data
+    structure type. Currently we support mutable sequences, mappings, and sets.
+
+    Args:
+        data:
+            Raw data to be potentially converted to a _HOOMDDataStructures
+            object.
+        parent:
+            The parent container if any for the data.
+        label:
+            Label to indicate to the parent where the child belongs. For
+            instance, if the parent is a mapping the label would be the key for
+            this value.
+    """
+    if isinstance(data, MutableMapping):
+        typing = _get_inner_typeconverter(type_def, TypeConverterMapping)
+        return HOOMDDict(typing, parent, data, label)
+    elif isinstance(data, MutableSequence):
+        typing = _get_inner_typeconverter(type_def, TypeConverterSequence)
+        return HOOMDList(typing, parent, data, label)
+    elif isinstance(data, MutableSet):
+        return HOOMDSet(typing, parent, data, label)
+    else:
+        return data
+
+
+class HOOMDList(MutableSequence, _HOOMDDataStructures):
+    """List with type validation.
+
+    Use `to_base` to get a plain `list`.
+
+    Warning:
+        Should not be instantiated by users.
+    """
+    def __init__(self, type_definition, parent, initial_value=None,
+                 label=None):
+        self._type_definition = type_definition
+        self._parent = parent
+        self._list = []
+        self._label = label
+        self._buffered = False
+        if initial_value is not None:
+            for type_def, val in zip(cycle(type_definition), initial_value):
+                self._list.append(_to_hoomd_data_structure(val, type_def, self))
+
+    def __getitem__(self, index):
+        """Get the value(s) associated with entry index of the list.
+
+        Supports slicing.
+        """
+        return self._list[index]
+
+    def __setitem__(self, index, value):
+        """Set an item(s) at index to the given value.
+
+        Support assignment with slices.
```

Comment:
```suggestion
        Supports assignment with slices.
```

Reply:
I don't know the exact use cases here but it's worth noting that NumPy-like data structures can also be assigned using a NumPy array of indices instead of a slice object. _If_ that is a useful feature, you could consider handling the case where the indices are a sequence (a list or array of indices).

Reply:
This is supposed to act like a standard list. We could support basic NumPy indexing, but I think that might be more confusing.